pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp

```diff
@@ -767,48 +767,42 @@ class StateVectorKokkos final
      * @param branch Branch 0 or 1.
      */
     void collapse(const std::size_t wire, const bool branch) {
-        auto &&num_qubits = this->getNumQubits();
-
-        const size_t stride = pow(2, num_qubits_ - (1 + wire));
-        const size_t vec_size = pow(2, num_qubits_);
-        const auto section_size = vec_size / stride;
-        const auto half_section_size = section_size / 2;
-
-        const size_t negbranch = branch ? 0 : 1;
-
-        Kokkos::MDRangePolicy<DoubleLoopRank> policy_2d(
-            {0, 0}, {half_section_size, stride});
+        KokkosVector matrix("gate_matrix", 4);
         Kokkos::parallel_for(
-            policy_2d,
-            collapseFunctor<fp_t>(*data_, num_qubits, stride, negbranch));
-
+            matrix.size(), KOKKOS_LAMBDA(const std::size_t k) {
```

Comment:
@vincentmr your method is much more elegant. Do you know if it makes any difference in terms of performance?

Reply:
If this were an operation used intensively and everywhere (e.g. CNOT) we should really opt for performance and benchmark it. I don't think it matters that much here, so I'm relying on intuition and propose changes for maintainability. That said, this implementation requires more flops and twice the reads, true, but that doesn't make it necessarily much slower. It might even come out faster depending on target because I've observed Kokkos' MDRangePolicy performance significantly drops for grids with extreme ratios (and  applyMultiQubitOp uses a simple RangePolicy which does not suffer that).